state/cleanup.go

```diff
@@ -1433,6 +1436,62 @@ func (st *State) cleanupForceRemoveMachine(machineId string, cleanupArgs []bson.
 	return machine.Remove()
 }
 
+// cleanupEvacuateMachine is initiated by machine.Destroy() to gracefully remove units
+// from the machine before then kicking off machine destroy.
+func (st *State) cleanupEvacuateMachine(machineId string, cleanupArgs []bson.Raw) error {
+	if len(cleanupArgs) > 0 {
+		return errors.Errorf("expected no arguments, got %d", len(cleanupArgs))
+	}
+
+	machine, err := st.Machine(machineId)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	if machine.Life() != Alive {
+		return nil
+	}
+
+	units, err := machine.Units()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(units) == 0 {
+		if err := machine.advanceLifecycle(Dying, false, false, 0); err != nil {
```

Comment:
can we double check all this works with `juju remove-machine --force`
that does invoke a separate code path - advanceLifecycle for non force is used here - but let's just be sure that `--force` works as expected when used

Reply:
Will test but pretty sure with force is unaffected.

Reply:
Force is unaffected by this.